File: moqt/internal/message/subscribe_update.go
package message

import (
	"io"
)

// SubscribeUpdateMessage updates the subscriber-side delivery preferences and
// group range for an existing SUBSCRIBE.
type SubscribeUpdateMessage struct {
	SubscriberPriority   uint8
	SubscriberOrdered    uint8
	SubscriberMaxLatency uint64
	StartGroup           uint64
	EndGroup             uint64
}

func (su SubscribeUpdateMessage) Len() int {
	var l int

	l += 1 // SubscriberPriority (uint8)
	l += 1 // SubscriberOrdered (uint8)
	l += VarintLen(su.SubscriberMaxLatency)
	l += VarintLen(su.StartGroup)
	l += VarintLen(su.EndGroup)

	return l
}

func (su SubscribeUpdateMessage) Encode(w io.Writer) error {
	var err error
	msgLen := su.Len()
	p := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	p, _, err = WriteMessageLength(p, uint64(msgLen))
	if err != nil {
		return err
	}
	p = append(p, su.SubscriberPriority)
	p = append(p, su.SubscriberOrdered)
	p, _, err = WriteVarint(p, su.SubscriberMaxLatency)
	if err != nil {
		return err
	}
	p, _, err = WriteVarint(p, su.StartGroup)
	if err != nil {
		return err
	}
	p, _, err = WriteVarint(p, su.EndGroup)
	if err != nil {
		return err
	}

	_, err = w.Write(p)

	return err
}

func (sum *SubscribeUpdateMessage) Decode(src io.Reader) error {
	size, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	if size > MaxMessageSize {
		return ErrMessageTooLarge
	}

	b := make([]byte, size)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	if len(b) < 2 {
		return ErrMessageTooShort
	}
	sum.SubscriberPriority = b[0]
	sum.SubscriberOrdered = b[1]
	b = b[2:]

	num, n, err := ReadVarint(b)
	if err != nil {
		return err
	}
	sum.SubscriberMaxLatency = num
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	sum.StartGroup = num
	b = b[n:]

	num, n, err = ReadVarint(b)
	if err != nil {
		return err
	}
	sum.EndGroup = num
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/subscribe_update.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageEncoder, parseUint8, parseVarint, readFull, readVarint } from "./message.ts";

export interface SubscribeUpdateMessageInit {
	subscriberPriority?: number;
	subscriberOrdered?: number;
	subscriberMaxLatency?: number;
	startGroup?: number;
	endGroup?: number;
}

export class SubscribeUpdateMessage {
	subscriberPriority: number;
	subscriberOrdered: number;
	subscriberMaxLatency: number;
	startGroup: number;
	endGroup: number;

	constructor(init: SubscribeUpdateMessageInit = {}) {
		this.subscriberPriority = init.subscriberPriority ?? 0;
		this.subscriberOrdered = init.subscriberOrdered ?? 0;
		this.subscriberMaxLatency = init.subscriberMaxLatency ?? 0;
		this.startGroup = init.startGroup ?? 0;
		this.endGroup = init.endGroup ?? 0;
	}

	/**
	 * Encodes the message to the writer.
	 */
	async encode(w: Writer): Promise<Error | undefined> {
		let buf: Uint8Array;
		try {
			const e = new MessageEncoder();
			e.uint8(this.subscriberPriority);
			e.uint8(this.subscriberOrdered);
			e.varint(this.subscriberMaxLatency);
			e.varint(this.startGroup);
			e.varint(this.endGroup);
			buf = e.frame();
		} catch (err) {
			return err as Error;
		}

		const [, err] = await w.write(buf);
		return err;
	}

	/**
	 * Decodes the message from the reader.
	 */
	async decode(r: Reader): Promise<Error | undefined> {
		let err: Error | undefined;

		let msgLen: number;
		[msgLen, , err] = await readVarint(r);
		if (err) return err;

		const buf = new Uint8Array(msgLen);
		[, err] = await readFull(r, buf);
		if (err) return err;

		let offset = 0;

		[this.subscriberPriority, offset] = (() => {
			const [val, n] = parseUint8(buf, offset);
			return [val, offset + n];
		})();

		[this.subscriberOrdered, offset] = (() => {
			const [val, n] = parseUint8(buf, offset);
			return [val, offset + n];
		})();

		[this.subscriberMaxLatency, offset] = (() => {
			const [val, n] = parseVarint(buf, offset);
			return [val, offset + n];
		})();

		[this.startGroup, offset] = (() => {
			const [val, n] = parseVarint(buf, offset);
			return [val, offset + n];
		})();

		[this.endGroup, offset] = (() => {
			const [val, n] = parseVarint(buf, offset);
			return [val, offset + n];
		})();

		return undefined;
	}
}
